libraries/ts-command-line/src/providers/TabCompletionAction.ts

```diff
@@ -0,0 +1,220 @@
+import { CommandLineStringParameter } from '../parameters/CommandLineStringParameter';
+import { CommandLineParameterKind, CommandLineParameter } from '../parameters/BaseClasses';
+import { CommandLineAction } from './CommandLineAction';
+import { CommandLineChoiceParameter } from '..';
+import { CommandLineConstants } from '../Constants';
+
+const DEFAULT_WORD_TO_AUTOCOMPLETE: string = '';
+const DEFAULT_POSITION: number = 0;
+
+export class TabCompleteAction extends CommandLineAction {
+  private _wordToCompleteParameter: CommandLineStringParameter;
+  private _positionParameter: CommandLineIntegerParameter;
+  private readonly _actions: Map<string, Map<string, CommandLineParameter>>;
+  private readonly _globalParameters: Map<string, CommandLineParameter>;
+
+  public constructor(
+    actions: ReadonlyArray<CommandLineAction>,
+    globalParameters: ReadonlyArray<CommandLineParameter>
+  ) {
+    super({
+      actionName: CommandLineConstants.TabCompletionActionName,
+      summary: 'Provides tab completion.',
+      documentation: 'Provides tab completion.'
+    });
+
+    this._actions = new Map<string, Map<string, CommandLineParameter>>();
+    for (const action of actions) {
+      const parameterNameToParameterInfoMap: Map<string, CommandLineParameter> = new Map<
+        string,
+        CommandLineParameter
+      >();
+      for (const parameter of action.parameters) {
+        parameterNameToParameterInfoMap.set(parameter.longName, parameter);
+        if (parameter.shortName) {
+          parameterNameToParameterInfoMap.set(parameter.shortName, parameter);
+        }
+      }
+      this._actions.set(action.actionName, parameterNameToParameterInfoMap);
+    }
+
+    this._globalParameters = new Map<string, CommandLineParameter>();
+    for (const parameter of globalParameters) {
+      this._globalParameters.set(parameter.longName, parameter);
+      if (parameter.shortName) {
+        this._globalParameters.set(parameter.shortName, parameter);
+      }
+    }
+  }
+
+  protected onDefineParameters(): void {
+    this._wordToCompleteParameter = this.defineStringParameter({
+      parameterLongName: '--word',
+      argumentName: 'WORD',
+      description: `The word to complete.`,
+      defaultValue: DEFAULT_WORD_TO_AUTOCOMPLETE
+    });
+
+    this._positionParameter = this.defineIntegerParameter({
+      parameterLongName: '--position',
+      argumentName: 'INDEX',
+      description: `The position in the word to be completed.`,
+      defaultValue: DEFAULT_POSITION
+    });
+  }
+
+  protected async onExecute(): Promise<void> {
+    const commandLine: string = this._wordToCompleteParameter.value || '';
+    const caretPosition: number = this._positionParameter.value || 0;
+
+    for await (const value of this.getCompletions(commandLine, caretPosition)) {
+      console.log(value);
+    }
+  }
+
+  public async *getCompletions(commandLine: string, caretPosition: number): AsyncIterable<string> {
```

Comment:
I don't understand why `getCompletions()` is async.  Isn't it a purely synchronous loop that simply enumerates all the CLI parameters?  The list of stuff that we are iterating is pretty short, like less than 1,000 lines of [CommandLineHelp.test.ts.snap](https://github.com/microsoft/rushstack/blob/master/apps/rush-lib/src/cli/test/__snapshots__/CommandLineHelp.test.ts.snap) snapshots.

(Disclosure: I find async code to be more complex and harder to debug, so I try to introduce it only where needed.)

Reply:
Also @sachinjoseph why doesn't your new `tab-complete` command show up in the snapshot?

Reply:
The test will now call parser.execute() which ensures that tab-completion is added the snapshot.